database/pgsql/pgsession.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019 clair authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pgsql
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/coreos/clair/database/pgsql/keyvalue"
+	"github.com/coreos/clair/database/pgsql/vulnerability"
+
+	"github.com/coreos/clair/database"
+	"github.com/coreos/clair/database/pgsql/ancestry"
+	"github.com/coreos/clair/database/pgsql/detector"
+	"github.com/coreos/clair/database/pgsql/feature"
+	"github.com/coreos/clair/database/pgsql/layer"
+	"github.com/coreos/clair/database/pgsql/lock"
+	"github.com/coreos/clair/database/pgsql/namespace"
+	"github.com/coreos/clair/database/pgsql/notification"
+	"github.com/coreos/clair/pkg/pagination"
+)
+
+type pgSession struct {
+	*sql.Tx
+
+	key pagination.Key
+}
+
+func (tx *pgSession) Commit() error {
+	return tx.Tx.Commit()
+}
+
+// Rollback drops changes to datastore.
+//
+// Rollback call after Commit does no-op.
+func (tx *pgSession) Rollback() error {
+	return tx.Tx.Rollback()
+}
+
+// UpsertAncestry inserts or replaces an ancestry and its namespaced
+// features and processors used to scan the ancestry.
+func (tx *pgSession) UpsertAncestry(a database.Ancestry) error {
+	return ancestry.UpsertAncestry(tx.Tx, a)
+}
+
+// FindAncestry retrieves an ancestry with all detected
+// namespaced features. If the ancestry is not found, return false.
+func (tx *pgSession) FindAncestry(name string) (database.Ancestry, bool, error) {
+	return ancestry.FindAncestry(tx.Tx, name)
+}
+
+// PersistDetector inserts a slice of detectors if not in the database.
+func (tx *pgSession) PersistDetectors(detectors []database.Detector) error {
+	return detector.PersistDetectors(tx.Tx, detectors)
+}
+
+// PersistFeatures inserts a set of features if not in the database.
+func (tx *pgSession) PersistFeatures(features []database.Feature) error {
+	return feature.PersistFeatures(tx.Tx, features)
+}
+
+// PersistNamespacedFeatures inserts a set of namespaced features if not in
+// the database.
+func (tx *pgSession) PersistNamespacedFeatures(features []database.NamespacedFeature) error {
+	return feature.PersistNamespacedFeatures(tx.Tx, features)
+}
+
+// CacheAffectedNamespacedFeatures relates the namespaced features with the
+// vulnerabilities affecting these features.
+//
+// NOTE(Sida): it's not necessary for every database implementation and so
+// this function may have a better home.
```

Comment:
This note should probably be on the Session interface rather than here.

Reply:
haha, it's on the session already. I'll remove all the comments.